### DISCUSSION TITLE
use @FastNative annotation for Android to speed up JNI calls

### DISCUSSION THREAD

**Discussion by kaitian521** [Tencent/MMKV#1149] [Ideas]
Let's talk about an interesting topic

[https://developer.android.com/reference/dalvik/annotation/optimization/FastNative](url)

========= my test ==========
```
        MMKV mmkv = MMKV.defaultMMKV();
        mmkv.encode("key", 1L);

        long start = System.currentTimeMillis();
        for (int i = 0; i < 1000000; i++) {
            long v = mmkv.decodeLong("key");
        }
        long duration = System.currentTimeMillis() - start;
        Log.e("lishaokai", "running duration = " + duration);
```

when adding @FastNative, running duration is 915ms
```
//    @FastNative
    private native long decodeLong(long handle, String key, long defaultValue);
```
when delete @fastnative, running duration is 1030ms

almost 10% speed up

so I just am curious about adding this annotation to speed up JNI calls for some special functions such as decodeX()
what's more, AOSP are using it, such as

https://github.com/aosp-mirror/platform_frameworks_base/blob/da436ad3e31f95eae922519ff1cbcf90b60900cb/core/java/android/os/Trace.java#L130

Thank you

**Comment by lingol** [comment]
Thanks for your input. I didn't know about that annotation until now.
